src/tests/ftest/slurm_setup.py

```diff
@@ -10,279 +10,481 @@
+        if not run_remote(self.log, self.control, command).passed:
+            raise SlurmSetupException(f'Error setting slurm epilog script {script} permissions')
+
+    def _copy_file(self, nodes, source, destination):
+        """Copy the source file to the destination on all the nodes.
+
+        Args:
+            nodes (NodeSet): nodes on which to copy the files
+            source (str): file to copy
+            destination (str): where to copy the file
+
+        Raises:
+            SlurmSetupException: if there is an error copying the file on any host
+        """
+        self.log.debug(f'Copying the {source} file to {destination} on {str(nodes)}')
+        command = command_as_user(f'cp {source} {destination}', self.root)
+        result = run_remote(self.log, nodes, command)
+        if not result.passed:
+            raise SlurmSetupException(
+                f'Error copying {source} to {destination} on {str(result.failed_hosts)}')
+
+    def _update_slurm_config(self, slurm_user, partition):
+        """Update the slurm config file.
+
+        Args:
+            slurm_user (str): user to define in the slurm config file
+            partition (str): name of the slurm partition to include in the configuration
+
+        Raises:
+            SlurmSetupException: if there is a problem modifying slurm config file
+        """
+        # Update the config file with the slurm cluster name
+        self._modify_slurm_config_file(
+            'slurm cluster name', self.all_nodes, 's/ClusterName=cluster/ClusterName=ci_cluster/g',
+            self.root)
+
+        # Update the config file with the slurm user
+        self._modify_slurm_config_file(
+            'slurm user', self.all_nodes, f's/SlurmUser=slurm/SlurmUser={slurm_user}/g',
+            self.root)
+
+        # Update the config file with the removal of the NodeName entry
+        self._modify_slurm_config_file(
+            'node name', self.all_nodes, 's/NodeName/#NodeName/g', self.root)
+
+        # Update the config file with the slurm epilog file
+        self._modify_slurm_config_file(
+            'epilog file', self.all_nodes, 's#EpilogSlurmctld=#EpilogSlurmctld={EPILOG_FILE}#g',
+            self.root)
+
+        # Update the config file with the slurm control node
+        not_updated = self.all_nodes.copy()
+        for control_keyword in ['SlurmctldHost', 'ControlMachine']:
+            command = f'grep {control_keyword} {self.SLURM_CONF}'
+            results = run_remote(self.log, self.all_nodes, command)
+            if results.passed_hosts:
+                not_updated.remove(
+                    self._modify_slurm_config_file(
+                        'slurm control node', results.passed_hosts,
+                        f's/{control_keyword}=linux0/{control_keyword}={str(self.control)}/g',
+                        self.root))
+        if not_updated:
+            raise SlurmSetupException(f'Slurm control node not updated on {not_updated}')
```

Comment:
We don't need to catch it because we want the `SlurmSetupException` exception raised and handled by the caller of `_update_slurm_config()`.

Reply:
In that case, does `_modify_slurm_config_file` even need to return the hosts that passed? Since either all hosts given passed, or an exception is raised